src/mds/MDSContext.h

```diff
@@ -135,11 +137,39 @@ class C_IO_Wrapper : public MDSIOContext
   C_IO_Wrapper(MDSRank *mds_, MDSInternalContextBase *wrapped_) : MDSIOContext(mds_), wrapped(wrapped_) {
     assert(wrapped != NULL);
   }
+
+  ~C_IO_Wrapper() {
+    delete wrapped;
+    wrapped = nullptr;
+  }
+
   virtual void finish(int r) {
     wrapped->complete(r);
   }
 };
 
+/**
+ * Like C_OnFinisher but cancellable, in which case we do
+ * cleanup manually instead of handling it to the finisher.
+ */
+class MDSIOFinisher: public Context {
```

Comment:
I don't think it's necessary to define a whole separate class: you can just add the deletion behavior to C_OnFinisher